Hooking backward for QNNPACK

kwanghoon-meta opened the pull request:
Summary: Enabling quantized gradient.

Test Plan:
Algorithmic correctness - Dequantized matmul vs QNNPACK matmul for gradient - P616202766

```
dequantized matmul : [1.5463, -0.2917, -2.1735, 0.5689, -1.0795]
QNNPACK matmul : tensor([[ 1.5463, -0.2917, -2.1735,  0.5689, -1.0795]])
```

Differential Revision: D42593235



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

albanD commented on torch/csrc/autograd/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,119 @@
+#include <ATen/native/quantized/PackedParams.h>
```
Could you move this to `torch/csrc/quantized/*` to make codeowners maintenance easier?

kwanghoon-meta replied:
alright let me change

albanD commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -58,6 +58,12 @@
 
+#if defined(__clang__)
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Werror"
```
ignoring warnings is not ok. We actually work hard to make this code warning free
cc @malfet

kimishpatel replied:
Why do you need to touch this file btw?

kwanghoon-meta replied:
Seems like my old code was creating a bug, which requires such change, now it does not.
Let me remove, thanks for a feedback.

albanD commented on buckbuild.bzl:
```diff
@@ -1463,6 +1463,7 @@ def define_buck_targets(
             "torch/csrc/jit/mobile/train/random.cpp",
             "torch/csrc/jit/mobile/train/sequential.cpp",
             ":gen_aten_libtorch[autograd/generated/Functions.cpp]",
+            "torch/csrc/quantized/quantized_backward.cpp",
```
I'm still surprised this is not in tools/build_variables.bzl. But if you only need it here it's ok. But that means that this won't be included in other targets and so you might get strange behavior when these backwards are not registered there.

kwanghoon-meta replied:
Hmmmm where should I place the code then?

albanD replied:
I'm not 100% sure tbh but I would have put it with similar files in https://github.com/pytorch/pytorch/blob/5b2ab0dd4f9b75657ba0e83641cd91864b4f23d1/build_variables.bzl#L156

albanD commented on torch/csrc/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,76 @@
+      DispatchKeySet({DispatchKey::CPU}), input, packed_weight, reduce_range);
+    auto [weight, bias] = packed_weight->unpack();
+    TORCH_CHECK(weight.qscheme() == c10::kPerTensorAffine);
+    // TO-DO : passing a bias does not work, hence create an arbitrary one in backward loop
```
If you need to save non-Tensor data, you can do so by doing `ctx->saved_data["foo"] = foo`

kimishpatel replied:
@albanD bias is also a tensor.

@kwanghoon-meta note that you are unpacking weights at forward, so how you will have two copies of weights. One in the packed_weight another for gradient that you are saving. So this will increase your memory footprint if that is what you wanted to minimize. Really what you want is to save packed_weight but I dont know if you can do that with autograd. But if this memory increase is fine then this works.

kimishpatel replied:
I think the reason passing bias dose not work is because gradient uses transposed weights where as bias shape is aligned for non-tranposed weight. But thinking more why do you need to pass bias at all or even use bias in the backward since gradient of bias addition w.r.t input is really zero, no?

kimishpatel replied:
> > Really what you want is to save packed_weight but I dont know if you can do that with autograd.
> 
> Just use `saved_data` ;)

oh thats right. You just said that above.

albanD commented on torch/csrc/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,76 @@
+};
+
+at::Tensor packed_linear_weight_grad(
+    at::Tensor input,
```
This should take `c10::DispatchKeySet ks, ` as the first argument. And this keyset should be used when redispatching during the forward pass. similar to https://github.com/pytorch/pytorch/blob/7a192cc51c7172860efd35413acf9a8b9aafd2c9/torch/csrc/autograd/VariableTypeManual.cpp#L110-L122

albanD commented on torch/csrc/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,76 @@
+    auto original_weight = saved[0];
+    original_weight = at::permute(original_weight, {1, 0});
+    auto grad_output = grad_outputs[0];
+    static auto op =
```
This is just calling a regular function, you can do it via `quantized::linear_prepack(original_weight, bias);`

The forward above is special because it is called while going down the dispatcher. So it needs to nicely redispatch.
This one is called by the autograd engine as regular code. So you can just do regular calls.

kimishpatel replied:
> The forward above is special because it is called while going down the dispatcher. So it needs to nicely redispatch.

@albanD what is the "nicely redispatch" part?

albanD replied:
> @albanD what is the "nicely redispatch" part?

Getting the `DispatchKeySet` argument from the caller and calling `redispatch` with the properly updated keyset.

Since you're not using codegen here, you indeed don't have a `quantized::` namespace. Then you can just update the line below to use `call()` instead of `redispatch()`.

kimishpatel commented on aten/src/ATen/native/cpu/CatKernel.cpp:
```diff
@@ -7,6 +7,14 @@
 #include <ATen/cpu/vec/vec.h>
 #include <c10/util/irange.h>
 
+// suppress warnings triggered by #pragma unroll:
```
Same for this one. If this has been working before I dont see why this diff has to introduce these changes.

kimishpatel commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6347,7 +6347,7 @@
   variants: method, function
   dispatch:
-    CPU, CUDA: zero_
+    CPU, CUDA, QuantizedCPU: zero_
```
@albanD do you know why this might be needed? As I commented on PR Zeroing out a float tensor is not the same as filling quantized tensor with just zeros.

albanD replied:
I don't know why this is needed no. Since there is no testing in this PR, I doubt any new codepath is actually exercised so I suspect this is not needed.
If you have specific test that requires this, I'm sure we can figure out why it is needed if it is indeed.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -65,13 +66,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
-        return std::tuple<at::Tensor, c10::optional<at::Tensor>>(original_tensor, bias_);
+        float* weight_scales_data = w_scales.data_ptr<float>();
+        at::Tensor weight_origin;
+        weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
```
Please use this API instead https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear.cpp#L240.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -65,13 +66,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
+        // See for the subtraction 128
+        // https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L319
+        weight_origin.sub_(128);
+        at::Tensor original_quantized_tensor = at::_make_per_tensor_quantized_tensor(weight_origin, weight_scales_data[0], w_zero_points[0]);
```
Also this is covering strictly per tensor path only, what about per channel. Either assert that the tensor being unpacked is per tensor quantized only or add path for per channel as well.

kwanghoon-meta replied:
This is very first step, we will expand to per channel accordingly. per channel requires more operation implementation in order to go through autograd engine (striding needs to be implemented etc)
Let me add a flag as of now with assert

kimishpatel replied:
Then at the least assert that the weight tensor is per channel.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -65,13 +66,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
+        // https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L319
+        weight_origin.sub_(128);
+        at::Tensor original_quantized_tensor = at::_make_per_tensor_quantized_tensor(weight_origin, weight_scales_data[0], w_zero_points[0]);
+        TORCH_CHECK(original_quantized_tensor.q_scale() == weight_scales_data[0]);
+        TORCH_CHECK(original_quantized_tensor.q_zero_point() == w_zero_points[0]);
```
Why this check? Shouldnt be needed.

malfet commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -58,6 +58,12 @@
 #define int_vector __m256i
 #endif // CPU_CAPABILITY_AVX512
 
+#if defined(__clang__)
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Werror"
```
Please use `C10_CLANG_DIAGNOSTIC_PUSH` and `C10_CLANG_DIAGNOSTIC_IGNORE`, but agree with @albanD that ignoring warnings require some explanation

malfet commented on aten/src/ATen/native/cpu/CatKernel.cpp:
```diff
@@ -7,6 +7,14 @@
 
+// suppress warnings triggered by #pragma unroll:
+// "warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]"
+#if defined(__clang__)
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Werror"
+#  pragma clang diagnostic ignored "-Wpass-failed"
+#endif
```
Use convenience macros and never ignore `-Werror` 
```suggestion
C10_CLANG_DIAGNOSTIC_PUSH()
C10_CLANG_DIAGNOSTIC_IGNORE("-Wpass-failed")
```

malfet reviewed (changes requested):
Please do not push `-Werror` to ignored settings, and if you have to push some exclusions use `C10_DIAGNOSTIC_PUSH_AND_IGNORED_IF_DEFINED`, which is backward compatible with older clangs where exclusion might not be defined

kimishpatel reviewed (changes requested):
Left some comments

malfet commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -65,13 +66,20 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
-        free(kernel);
-        return std::tuple<at::Tensor, c10::optional<at::Tensor>>(original_tensor, bias_);
+        float* weight_scales_data = w_scales.data_ptr<float>();
+        at::Tensor weight_origin;
+        weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
```
```suggestion
        auto weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
```

malfet commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -65,13 +66,20 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
+        float* weight_scales_data = w_scales.data_ptr<float>();
+        at::Tensor weight_origin;
+        weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
+        int8_t* weight_ptr_int8 =
+            reinterpret_cast<int8_t*>(weight_origin.data_ptr<int8_t>());
```
Why reinterpret cast is needed here?
```suggestion
        int8_t* weight_ptr_int8 = weight_origin.data_ptr<int8_t>();
```

malfet commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -65,13 +66,20 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
+        weight_origin.sub_(128);
+        // As of now, we are supporting only per tensor quantizer
+        // TO-DO : Support a per channel as well.
+        at::Tensor original_quantized_tensor = at::_make_per_tensor_quantized_tensor(weight_origin, weight_scales_data[0], w_zero_points[0]);
```
Nit
```suggestion
        auto original_quantized_tensor = at::_make_per_tensor_quantized_tensor(weight_origin, weight_scales_data[0], w_zero_points[0]);
```

malfet commented on torch/csrc/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,76 @@
+class PackedLinearWeightDynamicBackward :
+public Function<PackedLinearWeightDynamicBackward> {
+ public:
+  static torch::Tensor forward(
+    AutogradContext *ctx,
+    at::Tensor input,
```
What's the difference between `at::Tensor` and `torch::Tensor`? And why use `at::` namespace if you already have `using namespace at;` on line 7
```suggestion
  static Tensor forward(
    AutogradContext *ctx,
    Tensor input,
```

malfet reviewed (approved):
LGTM, but have a look at few of the nits

kwanghoon-meta committed:
Hooking backward for QNNPACK (#94432)

Summary:
Enabling quantized gradient.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/94432

Test Plan:
Algorithmic correctness - Dequantized matmul vs QNNPACK matmul for gradient - P616202766

```
dequantized matmul : [1.5463, -0.2917, -2.1735, 0.5689, -1.0795]
QNNPACK matmul : tensor([[ 1.5463, -0.2917, -2.1735,  0.5689, -1.0795]])
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Reviewed By: kimishpatel, malfet

Differential Revision: D42593235

Pulled By: kwanghoon-meta

fbshipit-source-id: 5744635f0cb90567e4bbf04b1318cfe35993f5e1
